frontend/public/components/monitoring.tsx

```diff
@@ -1068,21 +1122,85 @@ const SeriesButton = ({colorIndex, isDisabled, onClick}) => {
   ></button>;
 };
 
-const Query: React.FC<QueryProps> = ({colorOffset, onBlur, onDelete, onSubmit, onUpdate, query}) => {
-  const {allSeries, disabledSeries, enabled, expanded, text} = query;
+const QueryInput: React.FC<QueryInputProps> = ({metrics = [], onBlur, onSubmit, onUpdate, value = ''}) => {
+  const [token, setToken] = React.useState('');
 
-  const toggleEnabled = () => onUpdate({enabled: !enabled, expanded: !enabled, query: enabled ? '' : text});
+  const inputRef = React.useRef(null);
 
-  const toggleAllSeries = () => onUpdate({disabledSeries: _.isEmpty(disabledSeries) ? _.map(allSeries, 'labels') : []});
+  const onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+    onUpdate(e.target.value);
+
+    const textBeforeCursor = e.target.value.substring(0, inputRef.current.selectionEnd);
+
+    // Metric and function names can only contain the characters a-z, A-Z, 0-9, '_' and ':', but also add '(' and ')'
+    // for matching function brackets
+    const allTokens = textBeforeCursor.split(/[^a-zA-Z0-9_:()]+/);
+
+    setToken(_.last(allTokens));
+  };
 
   const onKeyDown = (e: React.KeyboardEvent) => {
     // Enter+Shift inserts newlines, Enter alone runs the query
     if (e.key === 'Enter' && !e.shiftKey) {
       e.preventDefault();
-      onSubmit(e);
+      onSubmit();
     }
   };
 
+  const onTextareaBlur = (e: React.FocusEvent<HTMLTextAreaElement>) => {
+    if (!_.get(e, 'relatedTarget.dataset.autocomplete')) {
+      onBlur(e);
+      setToken('');
+    }
+  };
+
+  const onClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
+    e.preventDefault();
+
+    // Replace the autocomplete token with the selected autocomplete option
+    const re = new RegExp(`${_.escapeRegExp(token)}$`);
+    onUpdate(value.replace(re, `${e.currentTarget.dataset.autocomplete} `));
+
+    setToken('');
+    inputRef.current.focus();
+  };
+
+  const functionOptions = token.length < 2 ? [] : prometheusFunctions.filter(a => fuzzy(token, a));
+  const metricOptions = token.length < 2 ? [] : metrics.filter(a => fuzzy(token, a));
+
+  return <div className="query-browser__query query-browser__inline-control pf-c-dropdown">
+    <textarea
+      autoFocus={true}
+      className="form-control query-browser__query-input"
+      onBlur={onTextareaBlur}
+      onChange={onChange}
+      onKeyDown={onKeyDown}
+      placeholder="Expression (press Shift+Enter for newlines)"
+      ref={inputRef}
+      value={value}
+    />
+    {(!_.isEmpty(functionOptions) || !_.isEmpty(metricOptions)) && (
+      <ul className="pf-c-dropdown__menu query-browser__metrics-dropdown-menu">
+        {!_.isEmpty(functionOptions) && <div className="text-muted query-browser__dropdown--subtitle">Functions</div>}
+        {_.map(functionOptions, f => <li key={f}>
+          <a href="#" className="pf-c-dropdown__menu-item" data-autocomplete={f} onClick={onClick}>{f}</a>
```

Comment:
nit: for accessibility, these would be better as `<button className="btn btn-link" type="button">`, but I'm not sure if that breaks the patternfly styles

Reply:
PR: https://github.com/openshift/console/pull/1996 (but without `btn btn-link` to get the correct styling).